### PR TITLE
[Splicing] Signer extended with method to sign prev funding transaction input

### LINKED ISSUE #3312: [Splicing] Signing utility for splicing
This is in preparation for Splicing (#1621 )

In splicing the channel has to sign the previous funding as an input to the new funding transactions. For this, the `EcdsaChannelSigner` trait has to be extended.

The proposed solution is visible in #3274, the `EcdsaChannelSigner` trait has been extended with this method:

```rust
	/// Create a signature for a splicing funding transaction, for the input which is the previous funding tx.
	fn sign_splicing_funding_input(
		&self, splicing_tx: &Transaction, splice_prev_funding_input_index: u16, splice_prev_funding_input_value: u64, redeem_script: &Script, secp_ctx: &Secp256k1<secp256k1::All>
	) -> Result<Signature, ()>;
```

with the implementation:

```rust
	fn sign_splicing_funding_input(&self, splicing_tx: &Transaction, splice_prev_funding_input_index: u16, splice_prev_funding_input_value: u64, redeem_script: &Script, secp_ctx: &Secp256k1<secp256k1::All>) -> Result<Signature, ()> {
		let sighash = &sighash::SighashCache::new(splicing_tx).segwit_signature_hash(splice_prev_funding_input_index as usize, &redeem_script, splice_prev_funding_input_value, EcdsaSighashType::All).unwrap()[..];
		let msg = hash_to_message!(sighash);
		let sig = sign(secp_ctx, &msg, &self.funding_key);
		Ok(sig)
	}
```

(files `src/sign/mod.rs` `src/sign/ecdsa.rs` `src/ln/channel.rs`)

### PR DESCRIPTION
**Pull request by optout21**
Fixes #3312 . (#1621 )

New method to sign an input of a transaction with our funding key.
Used for splicing, when signing the previous funding transaction. The previous funding transaction becomes an input to the new funding transaction, and it is a multisig, which we also need to sign.

### CODE REVIEW

File: lightning/src/sign/ecdsa.rs [hunk 1]
```diff
@@ -209,4 +209,12 @@ pub trait EcdsaChannelSigner: ChannelSigner {
 	fn sign_channel_announcement_with_funding_key(
 		&self, msg: &UnsignedChannelAnnouncement, secp_ctx: &Secp256k1<secp256k1::All>,
 	) -> Result<Signature, ()>;
+	/// Sign an input of a transaction with our funding key.
+	/// Used for splicing, when signing the previous funding transaction.
+	/// The previous funding transaction becomes an input to the new funding transaction,
+	/// and it is a multisig, which we also need to sign.
+	fn sign_transaction_input(
+		&self, tx: &Transaction, input_index: u16, input_value: u64, input_redeem_script: &Script,
```
[the remark this answers is not in the document]
Comment (optout21) [reply lightning/src/sign/ecdsa.rs:217]: Hm, I haven't thought about that.
The redeem script can be derived by the signer as well, true (funding pubkeys don't change).
The `input_index` is the index of the prev funding in the new funding (splicing) transaction, that's outside of the signer's scope.
The previous channel value could be used from the signer, unless already updated! The signing happens after the initial negotiation handshake, and in the current proto implementation (not set in stone) by that time the signer is already updated to the new channel value. However, I would keep this as a parameter, as it's more general, not to limit the space of possible implementations. (Another option would be to keep the whole previous signer.)
In conclusion, I remove the script parameter, but keep the value and index parameters.

File: lightning/src/sign/ecdsa.rs [hunk 2]
```diff
@@ -209,4 +209,12 @@ pub trait EcdsaChannelSigner: ChannelSigner {
 	fn sign_channel_announcement_with_funding_key(
 		&self, msg: &UnsignedChannelAnnouncement, secp_ctx: &Secp256k1<secp256k1::All>,
 	) -> Result<Signature, ()>;
+	/// Sign an input of a transaction with our funding key.
+	/// Used for splicing, when signing the previous funding transaction.
+	/// The previous funding transaction becomes an input to the new funding transaction,
+	/// and it is a multisig, which we also need to sign.
+	fn sign_transaction_input(
```
[the remark this answers is not in the document]
Comment (optout21) [reply lightning/src/sign/ecdsa.rs]: Renamed

File: lightning/src/sign/ecdsa.rs [hunk 3]
```diff
@@ -209,4 +209,12 @@ pub trait EcdsaChannelSigner: ChannelSigner {
 	fn sign_channel_announcement_with_funding_key(
 		&self, msg: &UnsignedChannelAnnouncement, secp_ctx: &Secp256k1<secp256k1::All>,
 	) -> Result<Signature, ()>;
+	/// Sign an input of a transaction with our funding key.
+	/// Used for splicing, when signing the previous funding transaction.
+	/// The previous funding transaction becomes an input to the new funding transaction,
```
[the remark this answers is not in the document]
Comment (optout21) [reply lightning/src/sign/ecdsa.rs:214]: Yes, I add more description for those parameters.

[the remark this answers is not in the document]
Comment (optout21) [reply lightning/src/sign/ecdsa.rs:214]: Technically the previous transaction can be unconfirmed, but I am not sure the spec allows for that. It is probably reasonable to exclude that. If a funding candidate will never confirm for some reason, creating an output on top of that will not confirm either.
From the point of view of this signer method the input must be an existing UTXO, and must be the same 2-of-2 multisig between the same public keys as the  current funding transaction.

File: lightning/src/sign/ecdsa.rs [hunk 4]
```diff
@@ -209,4 +209,15 @@ pub trait EcdsaChannelSigner: ChannelSigner {
...
+	/// and it is a multisig, which we also need to sign.
+	/// [`input_index`]: The index of the input that was the previous funding transaction,
```
Comment (jkczyz) [inline lightning/src/sign/ecdsa.rs:216]: I think this would be clearer as:

> The input within the new funding transaction `tx` spending the previous funding transaction's output.

File: lightning/src/sign/ecdsa.rs [hunk 5]
```diff
@@ -209,4 +209,15 @@ pub trait EcdsaChannelSigner: ChannelSigner {
...
+	///     within the new funding transaction.
+	/// [`input_value`]: The value of the previous funding transaction.
```
Comment (jkczyz) [inline lightning/src/sign/ecdsa.rs:218]: s/transaction/transaction output

File: lightning/src/sign/ecdsa.rs [hunk 6]
```diff
@@ -209,4 +209,15 @@ pub trait EcdsaChannelSigner: ChannelSigner {
 	fn sign_channel_announcement_with_funding_key(
 		&self, msg: &UnsignedChannelAnnouncement, secp_ctx: &Secp256k1<secp256k1::All>,
 	) -> Result<Signature, ()>;
+	/// Sign an input of a transaction with our funding key.
+	/// Used for splicing, when signing the previous funding transaction.
+	/// The previous funding transaction becomes an input to the new funding transaction,
+	/// and it is a multisig, which we also need to sign.
```
Comment (jkczyz) [inline lightning/src/sign/ecdsa.rs:215]: The wording here is a bit confusing because the first sentence says "sign and input" but the second sentence you're saying "signing the previous funding transaction". I think we can make this less verbose as:

> Signs the input of a splicing funding transaction with our funding key.
> 
> In splicing, the previous funding transaction output is spent as the input of the new funding transaction, and is a 2-of-2 multisig.

**Comment by optout21** [comment]
Documentation updated per the suggestions

File: lightning/src/sign/ecdsa.rs [hunk 7]
```diff
@@ -209,4 +209,18 @@ pub trait EcdsaChannelSigner: ChannelSigner {
...
+	///
+	/// In splicing, the previous funding transaction output is spent as the input of
+	/// the new funding transaction, and is a 2-of-2 multisig.
+	///
+	/// `input_index`: The index of the input within the new funding transaction `tx`,
+	///    spending the previous funding transaction's output
+	///
+	/// `input_value`: The value of the previous funding transaction output.
+	fn sign_splicing_funding_input(
+		&self, tx: &Transaction, input_index: u16, input_value: u64,
```
Comment (jkczyz) [inline lightning/src/sign/ecdsa.rs:223]: Could we use `usize` for `input_index` to be consistent with the rest of the API?

**Review by jkczyz — COMMENTED** [review]
LGTM modulo one small change

File: lightning/src/sign/ecdsa.rs (hunk 7 above)
Comment (optout21) [reply lightning/src/sign/ecdsa.rs]: Good point, done

**Comment by optout21** [comment]
> LGTM modulo one small change

Changed (also rebased)

**Review by jkczyz — APPROVED** [review]
